vault/identity_store.go

```diff
@@ -721,34 +721,48 @@ func (i *IdentityStore) invalidateEntityBucket(ctx context.Context, key string)
 				}
 			}
 
-			// If the entity is not in MemDB or if it is but differs from the
-			// state that's in the bucket storage entry, upsert it into MemDB.
-
 			// We've considered the use of github.com/google/go-cmp here,
 			// but opted for sticking with reflect.DeepEqual because go-cmp
 			// is intended for testing and is able to panic in some
 			// situations.
-			if memDBEntity == nil || !reflect.DeepEqual(memDBEntity, bucketEntity) {
-				// The entity is not in MemDB, it's a new entity. Add it to MemDB.
-				err = i.upsertEntityInTxn(ctx, txn, bucketEntity, nil, false)
-				if err != nil {
-					i.logger.Error("failed to update entity in MemDB", "entity_id", bucketEntity.ID, "error", err)
+			if memDBEntity != nil && reflect.DeepEqual(memDBEntity, bucketEntity) {
+				// No changes on this entity, move on to the next one.
+				continue
+			}
+
+			// If the entity exists in MemDB it must differ from the entity in
+			// the storage bucket because of above test. Blindly delete the
+			// current aliases associated with the MemDB entity. The correct set
+			// of aliases will be created in MemDB by the upsertEntityInTxn
+			// function. We need to do this because the upsertEntityInTxn
+			// function does not delete those aliases, it only creates missing
+			// ones.
+			if memDBEntity != nil {
+				if err := i.deleteAliasesInEntityInTxn(txn, memDBEntity, memDBEntity.Aliases); err != nil {
```

Comment:
@marcboudreau @elliesterner brought up a valid point while we were talking about entity merge prevention. Do you think not deleting the entity from memdb might cause an automatic merge to be triggered? if you could write a test for that, that would be awesome. we would like to prevent further merges from happening.

Reply:
From looking at the code in **(*IdentityStore).upsertEntityInTxn**, there are 2 circumstances that lead to **(*IdentityStore).mergeEntityAsPartOfUpsert** being called:
1. **previousEntity** is not nil and **entity** has an alias in its **Aliases** field that exists in MemDB and whose **CanonicalID** field is set to the value of `previousEntity.ID`
2. **entity** has an alias in its **Aliases** field that exists in MemDB and whose **CanonicalID** field is set to a value that is different than `entity.ID`.

In the **(*IdentityStore).invalidateEntityBucket** function, when **upsertEntityInTxn** is called, the **previousEntity** argument is always `nil`, so that rules out circumstance 1.

And by pre-deleting the aliases, we ensure that circumstance 2 cannot happen either.

Reply:
For the sake of clarity, pre-deleting the entity from MemDB and then calling `(*IdentityStore).upsertEntityInTxn` won't prevent an entity merge from happening, since the logic that decides that doesn't take into account whether the entity exists in MemDB or not. I think the only way to prevent an entity merge from happening, would be to scan each of the aliases associated with the entity (instead of pre-deleted them) and search for any alias in MemDB with a matching alias name and mount accessor and delete those. That would make it impossible for circumstance 2 to happen.